runtime/near-vm-logic/src/tests/miscs.rs

```diff
@@ -876,70 +876,194 @@ fn test_ed25519_verify() {
+            signature.clone(),
+            message.len(),
+            message.as_slice(),
+            public_key.len() - 1,
+            public_key.clone(),
+            Err(VMLogicError::HostError(HostError::Ed25519VerifyInvalidInput {
+                msg: "invalid public key length".to_string(),
+            })),
+        ),
+        (
+            bad_signature.len(),
+            bad_signature.clone(),
+            message.len(),
+            message.as_slice(),
+            public_key.len(),
+            public_key.clone(),
+            Ok(0),
+        ),
+        (
+            signature.len() - 1,
+            signature.clone(),
+            message.len(),
+            message.as_slice(),
+            public_key.len(),
+            public_key.clone(),
+            Err(VMLogicError::HostError(HostError::Ed25519VerifyInvalidInput {
+                msg: "invalid signature length".to_string(),
+            })),
+        ),
+        (
+            forged_signature.len(),
+            forged_signature.clone(),
+            message.len(),
+            message.as_slice(),
+            public_key.len(),
+            public_key.clone(),
+            Ok(0),
+        ),
+        (
+            forged_signature.len(),
+            forged_signature.clone(),
+            0,
+            message.as_slice(),
+            public_key.len(),
+            public_key.clone(),
+            Ok(0),
+        ),
+        (
+            signature.len(),
+            signature.clone(),
+            message.len(),
+            wrong_message_zero_bytes.as_slice(), // note that here the message len is correct - TODO: ensure that this is the desired behavior
+            public_key.len(),
+            public_key.clone(),
+            Ok(0),
+        ),
+    ];
 
-    let result = logic
-        .ed25519_verify(
-            bad_signature.len() as _,
-            bad_signature.as_ptr() as _,
-            message.len() as _,
+    for (
+        signature_len,
+        signature,
+        message_len,
+        message,
+        public_key_len,
+        public_key,
+        expected_result,
+    ) in scenarios
+    {
+        let result = logic.ed25519_verify(
+            signature_len as _,
+            signature.as_ptr() as _,
+            message_len as _,
             message.as_ptr() as _,
-            public_key.len() as _,
+            public_key_len as _,
             public_key.as_ptr() as _,
-        )
-        .unwrap();
+        );
+
+        assert_eq!(result, expected_result);
+        match expected_result {
+            Ok(1) | Ok(2) => {
+                assert_costs(map! {
+                    ExtCosts::read_memory_byte: 128,
+                    ExtCosts::read_memory_base: 3,
+                    ExtCosts::ed25519_verify_base: 1,
+                    ExtCosts::ed25519_verify_byte: 32,
+                });
+            }
+            // TODO: unsure how to verify the costs of other variants - would be nice to get a pointer
```

Comment:
I think the best way to do this is just explicitly pass expected costs with every test-case? 

I think that, at this point, rater htan writing a loop, its' better to add a helper fn and call it multiple times: 

```rust
#[track_caller]
fn check_ed25519_verify(
    signature_len: usize,
    signature: &[u8],
    message_len: usize,
    message: &[u8],
    public_key_len: usize,
    public_key: &[u8],
    want: Result<(), HostError>,
    want_costs: HashMap<ExtCosts, u64>,
) {
    ...
}

...

    check_ed25519_verify(
        signature.len(),
        &signature,
        message.len(),
        &message,
        public_key.len() - 1,
        &public_key,
        Err(VMLogicError::HostError(HostError::Ed25519VerifyInvalidInput {
            msg: "invalid public key length".to_string(),
        })),
        map! {
            ExtCosts::read_memory_byte: 128,
            ExtCosts::read_memory_base: 3,
            ExtCosts::ed25519_verify_base: 1,
            ExtCosts::ed25519_verify_byte: 32,
        },
    );

```